validation/validate_results.R

```diff
@@ -0,0 +1,160 @@
+        if (compare_values) {
+            ref_data <- RSQLite::dbGetQuery(ref_db_con, sprintf(table_all_data_query, table))
+            test_data <- RSQLite::dbGetQuery(test_db_con, sprintf(table_all_data_query, table))
+            diffs <- all.equal(ref_data[column_diffs$common], ref_data[column_diffs$common])
+            if (!isTRUE(diffs)){
+                logging::loginfo(
+                    sprintf(
+                        "%s - Differences found in values:\n%s",
+                        event_name, toString(diffs)
+                    )
+                )
+            }
+        }
+    }
+}
+
+
+compare_dfs <- function(reference, test, event_name=NA){
+    # 1. check for new columns
+    ref_columns <- names(reference)
+    test_columns <- names(test)
+    column_diff <- difference_between_lists(ref_columns, test_columns)
+
+    if (length(column_diff$reference_only) > 0) {
+        logging::loginfo(
+            sprintf(
+                "%s - Reference dataframe contains columns not found in test dataframe:\n%s",
+                event_name, toString(column_diff$reference_only)
+            )
+        )
+    }
+    if (length(column_diff$test_only) > 0) {
+        logging::loginfo(
+            sprintf(
+                "%s - Test dataframe contains new columns:\n%s",
+                event_name, toString(column_diff$test_only)
+            )
+        )
+    }
+
+    # 2. check for different values in existing columns
+    diffs <- all.equal(reference[column_diff$common], test[column_diff$common])
+    if (!isTRUE(diffs)){
+        logging::loginfo(
+            sprintf(
+                "%s - Differences found in column values:\n%s",
+                event_name, toString(diffs)
+            )
+        )
+    }
+    return(diffs)
+}
+
+
+if (length(data_dirs) > 0){
+    for (dir in data_dirs){
+        all_files_in_dir <- list.files(dir)
+        required_files_in_dir <- required_files %in% all_files_in_dir
+        if (all(required_files_in_dir)){
+            # check databases
+            ref_db_con <- RSQLite::dbConnect(RSQLite::SQLite(), sprintf("%s/%s", dir, ref_db_name))
+            test_db_con <- RSQLite::dbConnect(RSQLite::SQLite(), sprintf("%s/%s", dir, test_db_name))
+            compare_dbs(ref_db_con, test_db_con, TRUE, dir)
+
+            # check csvs
+            ref_circuit_summary <- read.csv(sprintf("%s/%s", dir, ref_circuit_summary_fname))
+            ref_underlying_data <- read.csv(sprintf("%s/%s", dir, ref_underlying_data_fname))
+            test_circuit_summary <- read.csv(sprintf("%s/%s", dir, test_circuit_summary_fname))
+            test_underlying_data <- read.csv(sprintf("%s/%s", dir, test_underlying_data_fname))
+
+            # check circuit summary
+            compare_dfs(ref_circuit_summary, test_circuit_summary, event_name=dir)
+            # check underlying data
+            compare_dfs(ref_underlying_data, test_underlying_data, event_name=dir)
+        }
+    }
+}
```

Comment:
Currently nothing is logged if there are no differences. It would be worth printing something so that the user knows it ran successfully

Reply:
Good thinking